lib/private/AppFramework/Utility/SimpleContainer.php

```
@@ -122,6 +124,32 @@ public function query(string $name, bool $autoload = true) {
 		throw new QueryException('Could not resolve ' . $name . '!');
 	}
 
+	public function injectFn(callable $fn) {
+		$reflected = new ReflectionFunction(Closure::fromCallable($fn));
+		return $fn(...array_map(function (ReflectionParameter $param) {
+			// First we try by type (more likely these days)
+			if (($type = $param->getType()) !== null) {
+				try {
+					return $this->query($type->getName());
+				} catch (QueryException $ex) {
+					// Ignore and try name as well
+				}
+			}
+			// Second we try by name (mostly for primitives)
+			try {
+				return $this->query($param->getName());
+			} catch (QueryException $ex) {
+				// As a last resort we pass `null` if allowed
+				if ($type !== null && $type->allowsNull()) {
+					return null;
+				}
+
+				// Nothing worked, time to bail out
+				throw $ex;
```

Should we log this as debug exception as well?

---

IMO we don't need to because the method is declare to throw this, so it's not unexpected and it would end in the logs anyway if not caught as we catch any `Throwable` when invoking `boot`.